docs(cn): translate content/docs/concurrent-mode-suspense.md into Chinese (#372)
reactjs/zh-hans.reactjs.org: content/docs/concurrent-mode-suspense.md

---
id: concurrent-mode-suspense

> title: Suspense 用于数据获取（试验版）

Reviewer: 和 @QC-L 讨论，统一叫试验阶段吧
```suggestion
title: 用于数据获取的 Suspense（试验阶段）
```

<div class="scary">

> > **注意**：

Reviewer 2: ```suggestion
>**注意**：
```

>

> > 本章节所描述的**实验性功能在稳定版本中[尚不可用](/docs/concurrent-mode-adoption.html)**。请不要在应用程序的生产环境中使用。这些功能可能会发生重大变化，并且在正式成为 React 的一部分之前不会给出警告。

Reviewer 2: ```suggestion
>本章节所描述的**实验性功能在稳定版本中[尚不可用](/docs/concurrent-mode-adoption.html)**。请不要在应用程序的生产环境中使用。这些功能可能会发生重大变化，并且在正式成为 React 的一部分之前不会给出警告。
```

>

> > 本文档面向早期此功能的使用者以及对此功能好奇的开发者。**如果你并不熟悉 React，请不必担心** —— 你不需要立刻学习这些功能。稳定版本中有相应的替代实现方式。举个例子，如果你在找和数据获取相关且当下可用的教程，可以改阅[这篇文档](https://www.robinwieruch.de/react-hooks-fetch-data/)。

Reviewer 2: ```suggestion
>本文档面向早期此功能的使用者以及对此功能好奇的开发者。**如果你并不熟悉 React，请不必担心** —— 你不需要立刻学习这些功能。稳定版本中有相应的替代实现方式。举个例子，如果你在找和数据获取相关且当下可用的教程，可以改阅[这篇文档](https://www.robinwieruch.de/react-hooks-fetch-data/)。
```

</div>

> React 16.6 新增了 `<Suspense>` 组件，让你可以“等待”目标代码的下载，并且可以直接指定一个加载的界面（像是个 spinner），让它在用户等待的时候显示：

Reviewer 2: ```suggestion
React 16.6 新增了 `<Suspense>` 组件，让你可以“等待”目标代码加载，并且可以直接指定一个加载的界面（像是个 spinner），让它在用户等待的时候显示：
```

```jsx
const ProfilePage = React.lazy(() => import('./ProfilePage')); // 懒加载

// 在 ProfilePage 组件处于加载阶段时显示一个 spinner
<Suspense fallback={<Spinner />}>
  <ProfilePage />
</Suspense>
```

> Suspense 用于数据获取，它是一个新特性，你可以使用 `<Suspense>` **以声明的方式来“等待”任何内容，包括数据。**本文重点介绍它在数据获取的用例，它也可以用于等待图像、脚本或其他异步的操作。

Reviewer: ```suggestion
用于数据获取的 Suspense 是一个新特性，你可以使用 `<Suspense>` **以声明的方式来“等待”任何内容，包括数据。**本文重点介绍它在数据获取的用例，它也可以用于等待图像、脚本或其他异步的操作。
```

> - [究竟什么是 Suspense ?](#what-is-suspense-exactly)

Reviewer 2: ```suggestion
- [究竟什么是 Suspense?](#what-is-suspense-exactly)
```

>   - [Suspense 不是什么](#what-suspense-is-not)

Reviewer 2: 我觉得这几个标题，可以讨论下

  - [Suspense 提供了什么](#what-suspense-lets-you-do)
- [在实践中使用 Suspense](#using-suspense-in-practice)
  - [如果我不用 Relay？](#what-if-i-dont-use-relay)
  - [致库作者](#for-library-authors)
- [传统实现方法 vs Suspense](#traditional-approaches-vs-suspense)
  - [方法 1：渲染之后获取数据（不使用 Suspense）](#approach-1-fetch-on-render-not-using-suspense)
  - [方法 2：接收到全部数据之后渲染（不使用 Suspense）](#approach-2-fetch-then-render-not-using-suspense)
  - [方法 3：获取数据之后渲染（使用 Suspense）](#approach-3-render-as-you-fetch-using-suspense)
- [尽早开始获取数据](#start-fetching-early)
  - [我们仍在寻求方法中](#were-still-figuring-this-out)
- [Suspense 和 Race Conditions](#suspense-and-race-conditions)
  - [涉及 useEffect 的 Race Conditions](#race-conditions-with-useeffect)
  - [涉及 componentDidUpdate 的 Race Conditions](#race-conditions-with-componentdidupdate)
  - [Race Conditions 问题](#the-problem)
  - [借助 Suspense 消除 Race Condition](#solving-race-conditions-with-suspense)
- [错误处理](#handling-errors)
- [下一步](#next-steps)

> ## 究竟什么是 Suspense？{#what-is-suspense-exactly}

Reviewer 2: ```suggestion
## 何为 Suspense？{#what-is-suspense-exactly}
```

> Suspense 让你的组件“等待”某个异步操作，直到该异步操作的结果可以被渲染。在下面[例子](https://codesandbox.io/s/frosty-hermann-bztrp)中，两个组件都在等待一个获取数据的异步 API 的返回：

Reviewer 2: ```suggestion
Suspense 让组件“等待”某个异步操作，直到该异步操作结束即可渲染。在下面[例子](https://codesandbox.io/s/frosty-hermann-bztrp)中，两个组件都会等待异步 API 的返回值：
```

}
```

> **[去 CodeSandbox 尝试](https://codesandbox.io/s/frosty-hermann-bztrp)**

Reviewer 2: ```suggestion
**[在 CodeSandbox 中尝试](https://codesandbox.io/s/frosty-hermann-bztrp)**
```

Reviewer 2: 后面同下

Author: 已将译文中十处codeSandbox跳转替换成建议版本～

> 上面的 demo 只是个示意。如果里头代码让你摸不着头脑，别担心。我们后面会详细说明这部分代码的运作方式。需要记住的是，Suspense 它其实更像是一个*机制*，而 demo 中那些具体的 APIs，像是`fetchProfileData()` 或者 `resource.posts.read()`，这些 APIs 本身并不重要。不过，如果你还是对它们很好奇，可以在这个 [demo sandbox](https://codesandbox.io/s/frosty-hermann-bztrp) 里头找到它们的定义。

Reviewer 2: ```suggestion
上面的 demo 只是个示意。如果示例代码让你摸不着头脑，别担心。我们后面会详细说明这部分代码的运作方式。需要记住的是，Suspense 其实更像是一种*机制*，而 demo 中那些具体的 API，如 `fetchProfileData()` 或者 `resource.posts.read()`，这些 API 本身并不重要。不过，如果你还是对它们很好奇，可以在这个 [demo sandbox](https://codesandbox.io/s/frosty-hermann-bztrp) 中找到它们的定义。
```

Reviewer 2: 里头太奇怪了...很口语化

Suspense 不是一个获取数据的库，而是一个机制。这个**机制是用来给获取数据的库**向 React 沟通说明*某个组件正在读取的数据当前仍不可用*。沟通之后，React 可以继续等待数据的返回并更新 UI。 在 Facebook，我们用了 Relay 和它那[支持 Suspense 的新功能](https://relay.dev/docs/en/experimental/step-by-step) 。我们期望其他的库，像 Apollo 之类的，也能支持 Suspense。

从长远上说，我们想让 Suspense 成为组件读取异步数据的主要方式——无论数据来自何方。

> ### Suspense 不是什么 {#what-suspense-is-not}

Reviewer 2: 这个翻译好奇怪

> Suspense 和当下其他解决异步问题的方法很不一样，因而，第一次接触 Suspense 容易让人产生误解。下面我们阐述下常见的误解：

Reviewer 2: ```suggestion
Suspense 和当下其他解决异步问题的方法有所区别，因而，第一次接触 Suspense 容易让人产生误解。下面我们阐述下常见的误解：
```

Author: 口吻区别不说，这里的”significantly“ 是有意弱化掉的吗？

Reviewer 2: 应该是译为有显著差别吧？
存在明显差异？

Reviewer 2: 我还是那个意见，很不一样太口语

> * **它不是数据获取的一种实现。**它并不假定你使用 GraphQL，REST，或者任何其他特定的数据格式、库、数据传输方式、协议。

Reviewer: 多删了一个空格，除了内容，其他地方和原文保持一致
```suggestion
*  **它不是数据获取的一种实现。**它并不假定你使用 GraphQL，REST，或者任何其他特定的数据格式、库、数据传输方式、协议。
```

> * **它不是数据获取的一种实现。**它并不假定你使用 GraphQL，REST，或者任何其他特定的数据格式、库、数据传输方式、协议。

Reviewer: 多删了一个空格，除了内容，其他地方和原文保持一致
```suggestion
 * **它不是数据获取的一种实现。**它并不假定你使用 GraphQL，REST，或者任何其他特定的数据格式、库、数据传输方式、协议。
```

* **它不是一个可以直接用于数据获取的客户端。**你不能用 Suspense 来“替代” `fetch` 或者 Relay。不过你可以使用集成 Suspense 的库（比如说，[新的 Relay API](https://relay.dev/docs/en/experimental/api-reference)）。

* **它不使数据获取与视图层代码耦合。**它协助编排加载状态在 UI 中的显示，但它并不将你的网络逻辑捆绑到 React 组件。

### Suspense 可以做什么 {#what-suspense-lets-you-do}

说了那么多，Suspense 到底有什么用呢？对于这个问题，我们可以从不同的角度来回答：

> - **它能让数据获取库与 React 紧密整合。**如果一个数据请求库实现了对 Suspense 的支持，那么，在 React 中使用 Suspense 将会是自然不过的事。

Reviewer: 原文使用了 `*` 不要改成 `-`

> - **它能让数据获取库与 React 紧密整合。**如果一个数据请求库实现了对 Suspense 的支持，那么，在 React 中使用 Suspense 将会是自然不过的事。

Reviewer 2: ```suggestion
* **它能让数据获取库与 React 紧密整合。**如果一个数据请求库实现了对 Suspense 的支持，那么，在 React 中使用 Suspense 将会是自然不过的事。
```

> - **它能让你有针对性地安排加载状态的展示。**虽然它不干涉数据_怎样_获取，但它可以让你对应用的视图加载顺序有更大的控制权。

Reviewer 2: ```suggestion
* **它能让你有针对性地安排加载状态的展示。**虽然它不干涉数据_怎样_获取，但它可以让你对应用的视图加载顺序有更大的控制权。
```

> - **它能够消除 race conditions。**即便是用上 `await`，异步代码还是很容易出错。相比之下，Suspense 更给人*同步*读取数据的感觉——假定数据已经加载完毕。

Reviewer: 这一块的格式和原文保持一致，从 108 - 122 行

Reviewer 2: 是不是用了格式化工具导致的？

> - **它能够消除 race conditions。**即便是用上 `await`，异步代码还是很容易出错。相比之下，Suspense 更给人*同步*读取数据的感觉 —— 假定数据已经加载完毕。

Reviewer 2: ```suggestion
* **它能够消除 race conditions。**即便是用上 `await`，异步代码还是很容易出错。相比之下，Suspense 更给人*同步*读取数据的感觉 —— 假定数据已经加载完毕。
```

## 在实践中使用 Suspense {#using-suspense-in-practice}

在 Facebook 中，我们目前只在生产环境使用整合了 Suspense 的 Relay。**如果当前你在找一份实质性的指引来上手 Suspense，[可以看 Relay 的这份指引](https://relay.dev/docs/en/experimental/step-by-step)**！指引里头写明了当前运行在我们在生产环境中的可用模式。

**上面的 demo 代码里用到的 API 其实是“假”的实现，不是 Relay**。我们这样做的目的是想让代码本身更易懂些，让不熟悉 GraphQL 的读者也能看懂代码。也正因为里头 API 是假的，demo 的代码本身并不是在应用中使用 Suspense 的“正确方式”。可以说，本文是从概念上出发，目的是帮你厘清*为什么* Suspense 是以特定方式运行，以及 Suspense 解决了哪些问题这两件事情。

### 如果我不用 Relay？{#what-if-i-dont-use-relay}

如果你当下并不使用 Relay，那么你暂时无法在应用中试用 Suspense。因为迄今为止，在实现了 Suspense 的库中，Relay 是我们唯一在生产环境测试过，且对它的运作有把握的一个库。

> 在接下来的几个月里，许许多多的库将会实现它们各自支持 Suspense 的 APIs。**如果你倾向于等到技术更加稳定之后才开始学习，那大概率你会先不看这部分文档，等到 Suspense 的生态更成熟之后再回来学习。**

Reviewer 3: ```suggestion
在接下来的几个月里，许多库将会实现它们各自支持 Suspense 的 API。**如果你倾向于等到技术更加稳定之后才开始学习，那大概率你会先不看这部分文档，等到 Suspense 的生态更成熟之后再回来学习。**
```

如果你有兴趣的话，也可以自己开发，然后将你对 Suspense 的实现整合到某个数据获取的库中。

### 致库作者 {#for-library-authors}

我们很期待看到社区中其他库对 Suspense 进行试验。对于数据获取库的作者，有一件重要的事情需要你们引起注意。

尽管实现对 Suspense 的支持从技术上是可行的，Suspense 当前**并不**作为在组件渲染的时候开始获取数据的方式。反而，它让组件表达出它们在正在“等待”*已经发出获取行为的*数据。**[使用 Concurrent 模式和 Suspense 来构建优秀的用户体验](/blog/2019/11/06/building-great-user-experiences-with-concurrent-mode-and-suspense.html)一文说明了这一点的重要性，以及如何在实践中实现这个模式。**

> 除非你有现成的解决方法来避免 waterfalls，我们建议采用支持在渲染之前就能先获取数据的 APIs。关于实现这类 API 的确切例子，你可以查看 [Relay Suspense API](https://relay.dev/docs/en/experimental/api-reference#usepreloadedquery) 实现预加载的方式。对于这方面的信息，我们当前给出的和过去给出的并不完全一致。因为 Suspense 用于数据获取还处于试验阶段，我们的推荐会随着我们对 Suspense 在生产环境中使用的习得和对 waterfalls 这个问题的理解深度不同，而发生变化。

Reviewer 3: ```suggestion
除非你有现成的解决方法来避免瀑布（waterfall）问题，我们建议采用支持在渲染之前就能先获取数据的 API。具体的例子，你可以查看 [Relay Suspense API](https://relay.dev/docs/en/experimental/api-reference#usepreloadedquery) 实现预加载的方式。过去我们在这里传达的信息并不完全一致。Suspense 用于数据获取还处于试验阶段，随着我们在生产环境中使用以及对问题的理解，我们的建议可能会随着时间的推移而变化。
```

## 传统实现方式 vs Suspense {#traditional-approaches-vs-suspense}

我们可以完全不提及当前主流的数据获取方式，只介绍 Suspense。但这样做的话，以下 3 件事情：Suspense 解决了的问题、为什么这些问题需要被处理、以及 Suspense 和其他现存方法的不同，会更难厘清。

> 因此，我们把 Suspense 看作是一系列解决方法的下一步逻辑演化。从这个角度对其展开介绍：

Reviewer 3: ```suggestion
取而代之的是，我们将 Suspense 视为一系列方法中的下一个逻辑步骤：
```

Author: “Suspense 视为方法的一个步骤”，这里是不是可以讨论下？我的理解中，它本身就是个独立的解决方法，而不是个方法中的某个步骤，是以往一系列解决方法的“进化”版。

> - **渲染之后获取数据（比如，在 useEffect 里获取数据）**：先开始渲染组件，然后每个完成渲染的组件可能会在它们的 effects 或者生命周期函数中获取数据。这种方式经常导致“waterfalls”。

Reviewer 3: ```suggestion
* **Fetch-on-render（如：在 `useEffect` 中 `fetch`）：**开始渲染组件。每个组件都有可能在它们的 effects 或者生命周期函数中触发数据的获取。这种方法经常导致“瀑布”问题。
```

> - **接收到全部数据之后渲染（比如，使用 Relay 但不使用 Suspense）**：先尽早获取下一屏需要的所有数据，当数据拿到之后，渲染新的屏幕。但在数据拿到之前，我们什么事也做不了。

Reviewer 3: ```suggestion
* ** Fetch-then-render（如：不使用 Suspense 的 Relay）：**尽早获取下一屏需要的所有数据，数据准备好后，渲染新的屏幕。但在数据拿到之前，我们什么事也做不了。
```

> - **获取数据之后渲染（比如，使用 Relay 且使用 Suspense）**：先尽早获取下一屏需要的所有数据，然后*马上* 着手渲染新的屏幕——*在我们接收到返回数据之前就开始*。一旦返回数据开始流入，React 再次渲染需要数据的那部分组件，直到数据全部接收完毕。

Reviewer 3: ```suggestion
* **Render-as-you-fetch（如：使用了 Suspense 的 Relay）：**尽早获取下一屏需要的所有数据，然后*立即*渲染新屏幕——*即在网络响应可用之前*。在接收到数据的过程中，React迭代地渲染需要数据的组件，直到渲染完所有内容为止。
```

>**注意**
>
>这部分流程经过了简化处理，在实际应用中，真正被采用的方法通常是由不同方法混合而成。但是，我们接下来还是会逐一审视这些方法，因为这样可以让我们更好理解它们各自的优劣。

为了对比这 3 个方法，我们分别用它们实现一个 profile 页面。

> ### 方法 1：渲染之后获取数据（不使用 Suspense）{#approach-1-fetch-on-render-not-using-suspense}

Reviewer 3: ```suggestion
### 方法 1：Fetch-on-Render（不使用 Suspense）{#approach-1-fetch-on-render-not-using-suspense}
```

目前 React 应用中常用的数据获取方式是使用 effect：

```js
// 在函数组件中：
useEffect(() => {
  fetchSomething();
}, []);

// 或者，在 class 组件里：
componentDidMount() {
  fetchSomething();
}
```

> 我们称这种方法为“渲染之后获取数据”（fetch-on-render），因为数据的获取是发生在组件被渲染到屏幕*之后*。这种方法会导致一个问题叫“waterfall”。

Reviewer 3: ```suggestion
我们称这种方法为“fetch-on-render”，因为数据的获取是发生在组件被渲染到屏幕*之后*。这种方法会导致一个名称为“瀑布”的问题。
```

}
```

**[在 CodeSandbox 中尝试](https://codesandbox.io/s/frosty-hermann-bztrp)**

以下是方法 3 中当我们渲染 `<ProfilePage>` 时会发生的事情：

1. 我们一开始就通过 `fetchProfileData()` 发出请求。这个方法返回给我们一个特殊的对象“resource”，而不是一个 Promise。在现实的案例中，这个对象是由像 Relay 集成 Suspense 来提供给我们的。
2. React 尝试渲染 `<ProfilePage>`。该组件返回两个子组件：`<ProfileDetails>` 和 `<ProfileTimeline>`。
3. React 尝试渲染 `<ProfileDetails>`。该组件调用了 `resource.user.read()`，但因为读取的数据还没被获取完毕，所以组件会处于一个“挂起”的状态。React 会跳过这个组件，转去渲染组件树中的其他组件。
4. React 尝试渲染 `<ProfileTimeline>`。该组件调用了 `resource.posts.read()`，和上面一样，数据还没获取完毕，所以这个组件也是处在“挂起”的状态。React 同样跳过这个组件，去渲染组件树中的其他组件。
5. 组件树中已经没有其他东西需要渲染了。因为 `<ProfileDetails>` 组件处于“挂起”状态，React 则是渲染出距该组件最近的上游 `<Suspense>` fallback 给它：`<h1>Loading profile...</h1>` 。渲染到这里就结束了。

这里的 `resource` 对象代表的数据虽然还没到位，但大概率它最终会被下载完。所以，当我们读取 `read()` 的时候，我们要么拿到数据，要么拿到一个处于“挂起”状态的组件。

**当返回数据开始流入的时候，React 会重新开始渲染，每一次渲染它都可能渲染出更加完整的组件树。**当 `resource.user` 的数据获取完毕之后，`<ProfileDetails>` 组件就能被顺利渲染出来，这时，我们就不再需要展示 `<h1>Loading profile...</h1>` 这个 fallback 了。当我们拿到全部数据之后，所有的 fallbacks 就都可以不展示了。

> 这个过程暗含着一个有意思的点，那就是，虽然我们是用 GraphQL 客户端来通过一个单一的请求来获取所有需要的数据，*但因为响应报文是数据流的格式，我们因此能够更早地展示出接收到的数据*。因为我们的采用的方法是“获取数据之后渲染”（render-as-we-fetch）（而非渲染之后才获取数据），我们能够在响应报文接收完毕之前就先“解锁”最外层的 `<Suspense>`。我们在方法 2 中没谈到这一点：即便是在方法 2 “接收到全部数据之后渲染”（fetch-then-render）中，在获取数据和渲染之间也有“瀑布”问题。而 Suspense 并不会导致“瀑布”，数据获取库像是 Relay 就抓住了 Suspense 的这个优势。

Reviewer 3: ```suggestion
这意味着一个有趣的事实，即使我们使用 GraphQL 客户端来收集单个请求中需要的所有数据，*流式响应也可以使我们尽早显示更多的内容*。*在数据获取时（as-we-fetch）*（而不是数据获取*后*）渲染，因此，如果 `user` 在响应中比 `posts` 出现得更早，我们则可以在响应结束之前“解锁”外层的 `<Suspense>` 边界。我们之前并没有意识到这一点，即便是 fetch-then-render（接收到全部数据之后渲染）这个解决方案，在数据获取和渲染之间也有“瀑布”问题。Suspense 没有这个“瀑布”问题，像 Relay 这样的库就利用了这个优势。
```

> 这里需要注意我们是如何在代码中去掉方法 2 中的 `if (…)` “is loading”这个检查分支。方法 3 不单单删去了 if 分支，还简化了代码设计快速转变的流程。举个例子，如果我们想让 `<ProfileDetails>` 组件和 `<ProfileTimeline>` 组件一直同时“弹出”，我们可以删去上面代码中的内层 `<Suspense>`。又或者，我们可以通过*分别给它们都包上一层*`<Suspense>` 来让两者的渲染展示独立于彼此。Suspense 赋予了我们对加载状态的精细控制力，让我们可以在不对代码进行大改的前提下，控制安排组件间的加载状态和显示顺序。

Reviewer 3: ```suggestion
请注意，我们是如何在组件中去掉 `if (…)`“is loading” 这个检查的。这不仅删除了样板代码，还简化了代码设计快速转变的流程。举个例子，如果我们想同时“弹出” `<ProfileDetails>` 组件和 `<ProfileTimeline>` 组件，只需删除两者之间的 `<Suspense>`。或者，我们可以通过*给它们各自的*`<Suspense>` 来让两者彼此独立。通过Suspense，我们可以更改加载状态的粒度并控制顺序，而无需调整代码。
```

## 尽早开始获取数据 {#start-fetching-early}

如果你正在开发获取数据的库，对于“获取数据之后渲染”这个方法，你会想知道一件至关重要的事情：**我们是在渲染*之前*就进行数据获取**。可以仔细观察下面代码：

```js
// 一早就开始数据获取，在渲染之前！
const resource = fetchProfileData();

// ...

function ProfileDetails() {
  // 尝试读取用户信息
  const user = resource.user.read();
  return <h1>{user.name}</h1>;
}
```

**[在 CodeSandbox 中尝试](https://codesandbox.io/s/frosty-hermann-bztrp)**

> 需要注意的是，上面代码中 `read()` 本身并不触发数据获取这个行为。它做的事情仅仅是**在数据获取之后**，去读取数据。这个区别对于用 Suspense 创建敏捷应用而言相当重要。我们并不想把数据获取推迟到组件渲染之后。因此，作为数据获取库的作者，你得实现这一点，让用户能拿到 `resource` 这个对象而不触发数据获取的行为。本文中所有的 demo 都通过使用“伪 API”来实现这点。

Reviewer 3: ```suggestion
请注意，此示例中的 `read()` 调用尚未*开始*。 它只是试图读取**已经获取的数据**。 在创建具有 Suspense 的快速应用程序时，这种差异非常重要。 我们不想在组件渲染之前延迟加载数据。 数据获取库作者可以通过确保在实际数据获取开始之前无法获取 `resource` 对象来强制实施此操作。 本文中所有使用“伪 API” 的演示都将强制执行此操作。
```

你可能认为演示代码中“在最顶层”获取数据的操作不切实际。如果我们想跳转到另一个人的 profile 页面怎么办？我们可能希望根据 props 获取数据。则解决问题的方法是：**在事件处理函数中开始获取数据**。下面是在不同 profile 页面间导航的简单示例：

```js{1,2,10,11}
// 开始获取数据，越快越好
const initialResource = fetchProfileData(0);

function App() {
  const [resource, setResource] = useState(initialResource);
  return (
    <>
      <button onClick={() => {
          const nextUserId = getNextId(resource.userId);
          // 再次获取数据：用户点击时
          setResource(fetchProfileData(nextUserId));

>         }}>

Reviewer: 原来的代码缩进不要改

}
```

**[在 CodeSandbox 中尝试](https://codesandbox.io/s/infallible-feather-xjtbu)**

通过上面代码中的方法，我们可以**并行获取代码和数据**。当我们在不同页面间跳转的时候，不需要等到页面对应的代码加载完之后才开始获取数据。我们可以实现在获取代码的同时（在点击链接的时候）也开始获取数据，从而提供更好的用户体验。

> 这个方法本身提出了一个问题：我们在渲染下个页面之前，怎么知道要获取*什么*数据？对于这个问题，有几个不同的解决方法（像是，通过将数据获取的相关代码整合到你的路由逻辑中来解决）。如果你在开发数据获取库，[使用 Concurrent 模式和 Suspense 来构建优秀的用户体验](/blog/2019/11/06/building-great-user-experiences-with-concurrent-mode-and-suspense.html)一文深入介绍了如何解决这个问题，以及问题本身的重要性。

Reviewer 3: ```suggestion
接下来出现的问题是如何知道在不渲染下一屏的情况下将获得什么。有几种方法可以解决此问题（例如，将数据请求集成到路由解决方案附近）。如果你正在开发数据请求库，那么[使用 Concurrent 模式和 Suspense 来构建优秀的用户体验](/blog/2019/11/06/building-great-user-experiences-with-concurrent-mode-and-suspense.html)将深入探讨如何实现此目标及其重要性。
```

### 我们仍在寻求方法中 {#were-still-figuring-this-out}

Suspense 本身作为一个机制而言，它灵活可变并且没有太多的限制。而产品的代码需要足够多的限制来保障代码中不会有“瀑布”问题。关于如何提供保障这一点，目前是有不同的实现方式。当下，我们仍在探索以下问题：

- 尽早地获取数据用代码表达起来会显得笨重。我们该如何让避免“瀑布”这个过程更加容易实现？
- 我们给某个页面获取数据时，当想快速地*从*该页面过渡切换出去，对应的 API 支持带上数据吗？
- 响应报文的有效时长是多少？缓存是要处理成全局还是本地？谁来操控相应的缓存？
- 可以不通过插入 `read()`，让代理协助表示懒加载的 APIs 吗？

> - 对于任意的 Suspense 数据，什么能作为复合 GraphQL queries 的等效替代？

Reviewer 3: ```suggestion
* 对于任意给定的 Suspense 数据，构建 GraphQL 查询看起来是什么样的？
```

> Relay 对于以上的问题有自己的一套答案。但上述问题的答案绝对不止一种，我们很期待看到 React 社区中解决这些问题的新思路。

Reviewer 3: ```suggestion
对于这些问题，Relay 有自己的答案。当然，实现这一目标的方法不止一种，我们很期待看到将出现在 React 社区的新想法。
```

> ## Suspense 和 竞争条件（race condition） {#suspense-and-race-conditions}

Reviewer 2: ```suggestion
## Suspense 和 Race Conditions {#suspense-and-race-conditions}
```

> Race conditions 是 bug 的一类，它的出现是源于开发人员对代码运行顺序的错误推算。在 `useEffect` Hook 或在 class 组件的生命周期函数中调用 `componentDidUpdate` 方法是导致 race conditions 的常见原因。对这类问题，Suspense 也能帮得上忙——下面谈谈具体做法。

Reviewer: 如果要括号注释，就在全文前文首次出现的地方标注吧

> 为了说明这类问题，我们增加一个最外层组件 `<App>`，它渲染出我们上面的`<ProfilePage>` 组件，外加一个按钮，让我们可以**在不同的 profiles 页面之间切换**。

Reviewer 3: ```suggestion
为了说明问题，我们增加一个顶层组件 `<App>` 来渲染 `<ProfilePage>` ，并放置一个可以**在不同的 profile 页面之间切换**的按钮：
```

}
```

> 接下来，我们对比下不同的数据获取方法分别是如何实现页面切换。

Reviewer 3: ```suggestion
让我们比较一下不同的数据获取方法如何实现这个需求。
```

> ### 涉及 `useEffect` 的 Race Conditions {#race-conditions-with-useeffect}

Reviewer 3: ```suggestion
### 竞争条件（race condition）与 `useEffect` {#race-conditions-with-useeffect}
```

> 我们先直接复制上面方法 1 的代码，然后做些修改：给 `<ProfilePage>` 组件传入个参数 `id`，并把这个 `id` 传给 `fetchUser(id)` 和 `fetchPosts(id)`，如下：

Reviewer 3: ```suggestion
首先，我们将尝试使用原始的“在 effect 中获取数据”示例，重写在 `<ProfilePage>` props 中传递的 `id` 参数，以传给 `fetchUser(id)` 和 `fetchPosts(id)`：
```

}
```

**[在 CodeSandbox 中尝试](https://codesandbox.io/s/nervous-glade-b5sel)**

> 需要注意代码中 effect 的依赖从 `[]` 变成了 `[id]`——因为我们想在 `id` 变化之后，effect 紧接着再次运行，不然的话，我们就拿不到最新的数据。

Reviewer 3: ```suggestion
请注意，我们将 effect 的依赖从 `[]` 更改为 `[id]`——因为我们希望每次 `id` 更改时都重新运行 effect。 如果不这样做，我们将无法再次获取到新的数据。
```

> 如果我们运行上面的代码，咋一看会觉得它应该可以正常运行。然而，如果我们在 `fetchUser` 和 `fetchPosts` 这两个“伪 API”里头都做延迟处理，且延迟的时间随机赋值，接着快速点击按钮“Next”，我们就能从 console 的打印信息看出程序有 bug。**在我们已经切换到新的 profile 页面之后，旧页面发出的请求会时不时“杀回来”——那时，回来的过期响应报文就会用另外一个 ID 的过期数据重写当前正确且新鲜的 state。**

Reviewer 3: ```suggestion
如果我们尝试运行此代码，一开始看起来运行得很好。但是，如果我们将“伪 API”实现的延迟时间随机化，快速点击“Next”按钮，就会发现控制台日志中有些问题。**有时，在把 profile 页面切换成别的 ID 后，旧的 profile 的请求会“返回”——这种情况下，其他 ID 用的旧响应会覆盖新的 state。**
```

  const [resource, setResource] = useState(initialResource);
```

当我们点击“Next”按钮，`<App>` 组件便开始发出获取下一个 profile 的请求，并把*请求返回的*对象下传给 `<ProfilePage>` 组件。

```js{4,8}
<>
  <button onClick={() => {
      const nextUserId = getNextId(resource.userId);
      setResource(fetchProfileData(nextUserId));

>     }}>

Reviewer: 原来的代码缩进不要改

}
```

**[在 CodeSandbox 中尝试](https://codesandbox.io/s/adoring-goodall-8wbn7)**

上面代码中的错误边界组件既能捕捉渲染过程的报错，*也*能捕捉 Suspense 里头数据获取的报错。理论上，我们在组件树中插入多少个错误边界组件都是可以的，但这并不是推荐的做法，错误边界组件的位置最好是深思熟虑之后再确定。

## 下一步 {#next-steps}

到这里，我们已经介绍完当 Suspense 用于数据获取时的基本内容。更重要的是，我们现在对 Suspense 有自己运作方式的*原因*，以及它是如何在数据获取这个领域中发挥自己的作用这两点有更好的理解。

Suspense 本身解答了一些问题，但同时它也引出一些新的问题：

- 如果部分组件处于“挂起”状态，整个应用会卡死吗？该怎么避免这个问题？
- 如果我们不想在目标组件的上层，而想在其他地方展示 spinner，可以实现吗？
- 如果我们*想*有计划地在一个短的时间内展示不同的 UI，能够实现吗？
- 除了展示个 spinner，我们能添加额外的视觉效果吗？像是给现有界面加上蒙层之类的？

> - 在[最后一个 Suspense 的代码示例中](https://codesandbox.io/s/infallible-feather-xjtbu)，为什么在点击了“Next”按钮之后，会报出警告？

Reviewer: md 符号和原文保持一致
```suggestion
* 如果部分组件处于“挂起”状态，整个应用会卡死吗？该怎么避免这个问题？
* 如果我们不想在目标组件的上层，而想在其他地方展示 spinner，可以实现吗？
* 如果我们*想*有计划地在一个短的时间内展示不同的 UI，能够实现吗？
* 除了展示个 spinner，我们能添加额外的视觉效果吗？像是给现有界面加上蒙层之类的？
* 在[最后一个 Suspense 的代码示例中](https://codesandbox.io/s/infallible-feather-xjtbu)，为什么在点击了“Next”按钮之后，会报出警告？
```
